### PR TITLE
Upgrade Linkedin API version from 202304 to 202403

### PR DESCRIPTION
**Pull request by sgandhi1311**
# Description of change
- **API Updates**
  - Bumped to API version `202304` ([#64](https://github.com/singer-io/tap-linkedin-ads/pull/69))
  - Updated API endpoints for the following streams: `campaign_groups`, `campaigns`, `creatives`

- **Pagination Enhancements**
  - Implemented cursor-based pagination for the following streams: `accounts`, `campaign_groups`, `campaigns`, `creatives`

- **API Query Param Adjustments**
  - Removed unsupported `pivot` and `pivotValue` from `fields` query parameters in Analytics API requests
  - Incorporated these values within the tap to ensure consistency with the previous tap version

- **Video Ads Stream**
  - Added new fields ([#71](https://github.com/singer-io/tap-linkedin-ads/pull/71))
  - Requires scope - `r_organization_social` to sync the records.
  - data not available for `change_audit_stamps` in new endpoint

# Manual QA Steps

- **Stream Verification**
  - Synced all streams and verified the updated API endpoints for correctness.
  
- **Pagination Testing**
  - Conducted tests to ensure the pagination logic functions correctly.
  
- **Data Consistency Check**
  - Verified the record count aligns with the previous tap version 2.2.0.
  
- **Test Suite Maintenance**
  - Updated and fixed unit tests and integration tests to reflect the recent changes.

# Risks

- **Risk Level**: Medium
  - Despite thorough testing, there is a possibility of encountering corner case scenarios.
  - The changes were made quickly to meet the API's deadline before it expired.

 
# Rollback steps
 - Fix the error (if any) and deploy the change, as rollback won't work due to the API expiry.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,23 @@
...
+- **Pagination Enhancements**
+  - Implemented cursor-based pagination for the following streams: `accounts`, `campaign_groups`, `campaigns`, `creatives`
+
+- **API Query Param Adjustments**
+  - Removed unsupported `pivot` and `pivotValue` from `fields` query parameters in Analytics API requests
+  - Incorporated these values within the tap to ensure consistency with the previous tap version
+
+- **Video Ads Stream**
+  - Added new fields ([#71](https://github.com/singer-io/tap-linkedin-ads/pull/71))
+  - Requires scope - `r_organization_social` to sync the records.
```
Comment (rdeshmukh15) [inline CHANGELOG.md:19]: Scope requirements should be mentioned in README.md and not in CHANGELOG.md unless you have made changes on tap side specific to the scope.

File: tap_linkedin_ads/streams.py [hunk 2]
```diff
@@ -575,12 +612,25 @@ class VideoAds(LinkedInAds):
...
+        "dscAdTypes": "List(VIDEO)"
     }
+    headers = {'X-Restli-Protocol-Version': "2.0.0"}
+
+    def sync_endpoint(self, *args, **kwargs):
+        try:
+            return super().sync_endpoint(*args, **kwargs)
+        except Exception as error:
+            if "Not enough permissions to access: partnerApiPostsExternal" in str(error):
+                LOGGER.info("Access to the video-ads API is denied due to insufficient permissions. Please reauthenticate or verify the required permissions.")
```
Comment (rdeshmukh15) [inline tap_linkedin_ads/streams.py:629]: Using LOGGER.warning seems more suitable at the place of LOGGER.info

File: tap_linkedin_ads/streams.py [hunk 3]
```diff
@@ -561,10 +599,9 @@ class Accounts(LinkedInAds):
     data_key = "elements"
     children = ["video_ads"]
     params = {
-        "q": "search",
-        "sort.field": "ID",
-        "sort.order": "ASCENDING"
+        "q": "search"
```
Comment (rdeshmukh15) [inline tap_linkedin_ads/streams.py:602]: Is the data coming in sorted order, by default, in the new API version?

Comment (sgandhi1311) [reply tap_linkedin_ads/streams.py:602]: yes, as mentioned in [docs](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-accounts?view=li-lms-2024-03&tabs=http#search-parameters)

File: tap_linkedin_ads/streams.py [hunk 4]
```diff
@@ -129,7 +147,10 @@ def merge_responses(data):
         # Loop through each record of the page
         for element in page:
             temp_start = element['dateRange']['start']
-            temp_pivotValue = element['pivotValue']
+            temp_pivotValue = element['pivotValues'][0]
+            # adding pivot and pivot_value to make it compatible with the previous tap version 2.2.0
+            element['pivot'] = pivot
+            element["pivot_value"] = temp_pivotValue
```
Comment (rdeshmukh15) [inline tap_linkedin_ads/streams.py:153]: Making tap compatible with the previous version 2.2.0 is really needed? Because the previous tap version is already deprecated.

**Review by rdeshmukh15 — CHANGES_REQUESTED** [review]
Not a blocker to merge this PR but should be considered removing the pivot and pivotValue occurrences if it is not supported anymore.

File: CHANGELOG.md (hunk 1 above)
Comment (sgandhi1311) [reply CHANGELOG.md]: done

File: tap_linkedin_ads/streams.py (hunk 2 above)
Comment (sgandhi1311) [reply tap_linkedin_ads/streams.py]: done

File: tap_linkedin_ads/streams.py (hunk 4 above)
Comment (sgandhi1311) [reply tap_linkedin_ads/streams.py]: removed 2.2.0 from comments

**Comment by sgandhi1311** [comment]
> Not a blocker to merge this PR but should be considered removing the pivot and pivotValue occurrences if it is not supported anymore.

We can think of it in the future, currently, the fields - pivot and pivotValue are set to automatic. This PR targets to update the API version on priority.

**Review by Vi6hal — APPROVED** [review]

**Comment by Vi6hal** [comment]
@rdeshmukh15 there are few more deprecations with video_ads stream too, but since the api deprecation deadline is very near, we don't want to remove any schema objects thus avoiding a major version bump.

i am going to document these compatibility changes to be done in the next major release.
